blockprod/src/detail/block_maker.rs

```diff
@@ -185,3 +185,460 @@ impl BlockMaker {
         Ok(())
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use crate::tests::setup_blockprod_test;
+    use chainstate_types::{BlockIndex, BlockPreconnectData, ConsensusExtraData};
+    use crypto::random::make_pseudo_rng;
+    use mempool::{MempoolInterfaceMock, MempoolSubsystemInterface};
+
+    use chainstate::{
+        chainstate_interface::ChainstateInterface,
+        BlockError::{self, PrevBlockNotFound},
+        {ChainstateError, MockChainstateInterfaceMock},
+    };
+
+    use common::{
+        chain::{
+            block::{timestamp::BlockTimestamp, BlockReward, ConsensusData},
+            Block,
+        },
+        primitives::{BlockHeight, Id, H256},
+        time_getter::TimeGetter,
+        Uint256,
+    };
+
+    use super::*;
+
+    #[tokio::test]
+    async fn collect_transactions_subsystem_error() {
+        let (mut manager, chain_config, chainstate, _mempool) = setup_blockprod_test();
+
+        let mock_mempool = MempoolInterfaceMock::new();
+
+        let mock_mempool_subsystem = manager.add_subsystem_with_custom_eventloop("mock-mempool", {
+            let mock_mempool = mock_mempool.clone();
+            move |call, shutdn| async move {
+                mock_mempool.run(call, shutdn).await;
+            }
+        });
+
+        mock_mempool_subsystem.call_mut({
+            let shutdown = manager.make_shutdown_trigger();
+            move |_| shutdown.initiate()
+        });
+
+        // shutdown straight after startup, *then* call collect_transactions()
+        manager.main().await;
+
+        let (_tx_builder, rx_builder) = crossbeam_channel::unbounded();
+
+        tokio::spawn(async move {
```

Comment:
You can make the tester task into a subsystem with a custom event loop (which does not actually contain a loop). When it exits, the subsystem manager will be automatically notified and will tear down the remaining subsystems in a clean way, no custom shutdown handling required.

Reply:
Of course! Ok cool, I'll change this which should make things simpler.

Reply:
Apart from the tests that have to run after `master.main()` is called, all of the `spawn()` calls have now been converted.